### DISCUSSION TITLE
FadeCandy alternative?

### DISCUSSION THREAD

**Discussion by Pucksnok** [missionpinball/missionpinball#56] [General]
Hi. I'm in a huge need for more LED's. Is there any alternative to FadeCandy?  Seems it's not available to buy since a year back or something.
Preferably something that doesn't cost me a lot...

https://docs.missionpinball.org/en/stable/hardware/fadecandy/index.html

**Comment by cobra18t** [comment]
What about an OPP STM32 board? It costs less than a Fadecandy and supports 256 LEDs in a single chain. Not the same as the 64x8 LEDs that the Fadecandy does, but similar. It is a 3.3V device, but as long as you place it close to the first LED, you should be fine. I add a 3.3V->5V level converter to my boards because I want to remove that variable for my customers. You just use B15 as the data line as shown here: https://pinballmakers.com/wiki/index.php?title=OPP#STM32F103
SPI LEDs are also possible, but I forget the CLK pin # at the moment.

**Reply by Pucksnok** [reply]
Thx, seems a bit more complex than I hoped but ordered a few and a programmer for them, can figure it out later, I usually do in some magic way. :D

As I understand it I can also use the board for switches without any extra "wing PCBs". 

If I want to use a few of them I must chain them or can I plug them all straight into usb on computer?

Will this create any problems when using my cobra boards?

**Reply by Pucksnok** [reply]
To answer myself on one part I shouldn't need to chain the cards. I can use several chains so I should be able to connect them all trough USB.

https://docs.missionpinball.org/en/latest/hardware/opp/connecting.html

Now I just need to understand how every switch and the LEDs are named and if it could interfere with the current cobrapin boards I have setup.

**Reply by cobra18t** [reply]
Yes, just connect via USB and use B15 for Data and the GND. Supply your 5V with a shared GND from elsewhere.

As far as interfering with the other boards, you should set the serial number to something other than 0-4 since I use those for CobraPin. You can bypass the serial number business if you directly assign ports, but this can lead to the boards getting mixed up, so I do not recommend it.

You can always buy the boards from me on Pinside and tell me what you want and I will load the firmware and configuration including the serial number. The boards I have already have pin headers which you may or may not want. I will add a variant to the listing that just includes LED output, switch inputs, and a serial number of 252 (0xFC for FadeCandy). Then you would address the LEDs from 252-0-0 to 252-0-255.

You can use the other pins as inputs, just understand that there is no protection and they are only 3.3V tolerant. Here is more info on MPF setup of switches:
https://docs.missionpinball.org/en/latest/hardware/opp/switches.html

**Reply by Pucksnok** [reply]
Thanks, will do a order as soon the SEK currency is a bit stronger, been waiting for a while so can continue waiting a bit longer. Will order a satellite pcb and some extra processor-cards if something goes wrong with the ones I have. 

I now have two STM32 running and as standard it sets it as switch wings, do I have to use any other config to use neoled?
Not really understanding how the "config".py-files are built, I had no problem to set a serial and erase/write but seems I need to learn how the files must be built. Read the docs in the opp source but doesn't seem to cover it.

![image](https://user-images.githubusercontent.com/84498221/215597457-aa6ebaf1-15f7-465c-9e22-448ce4ea4e74.png)

**Reply by Pucksnok** [reply]
I guess something like this is correct. Copy paste from different configuration-files.
Wing 0 Neo, rest input

```

testVers = '00.00.02'


import rs232Intf


# Config inputs as all state inputs

wingCfg = [ [ rs232Intf.WING_NEO, rs232Intf.WING_INP, rs232Intf.WING_INP, rs232Intf.WING_INP ] ]


# Config inputs as all state inputs

inpCfg = [ [ rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, \

             rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, \

             rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, \

             rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, \

             rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, \

             rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, \

             rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, \

             rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE, rs232Intf.CFG_INP_STATE ] ]



# Config color table

# bytesPerPxl (1 byte), numPixels (1 byte, if 0, then 256 pixels), unused (95 bytes) 

colorCfg = [ [ '\x03', '\x08', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', \

               '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', \

               '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', \

               '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', \

               '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', \

               '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', \

               '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', \

               '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', '\xff', \

               '\xff', \es

```

**Reply by Pucksnok** [reply]
And while I ask stuff. To update my cobra pin processors is it just to flash a new opp firmware to get them updated? Serial is kept as long I don't erase the stm32, is the config saved as well or do I have to reload them?

**Reply by cobra18t** [reply]
Here is my OPP config: https://github.com/cobra18t/CobraPin/blob/main/OPP%20Config/CobraPin_Board252_FC_inputs.py

It enables 256 LEDs (yours only enables 8) and defaults them to a dim blue glow for troubleshooting purposes (yours defaults to a red, green, blue, white pattern that is quite bright and power hungry for a test mode)

I always flash new firmware, then reload the serial number and the OPP config just to be sure.

**Reply by Pucksnok** [reply]
I know I didn't know what I did with the config. Thanks for sharing!!
Will update my Cobrapin boards tonight with new opp firmware, should be easy enough. Again, thank you!

**Comment by Pucksnok** [comment]
I can now after all the help in here say that STM32F103C8T6 works perfectly for neopixels with everything I thrown at it. I now have two STM32 I tried and after setting a serial ("-ser=xxx") they work perfectly together.
Thanks for all the help.
